Mathlib/RingTheory/Nilpotent/GeometricallyReduced.lean

```diff
@@ -0,0 +1,144 @@
+  are geometrically reduced, then `A` is geometrically reduced.
+
+## References
+- See [https://stacks.math.columbia.edu/tag/05DS] for some theory of geometrically reduced algebras.
+  Note that their definition differs from the one here, we still need a proof that these are
+  equivalent.
+
+-/
+
+noncomputable section
+variable (k : Type) (A : Type) [Field k] [CommRing A] [Algebra k A]
+variable (B : Type) [CommRing B] [Algebra k B]
+
+
+/-- Given a morphism f : A →ₐ[k] B of k-algebras and a k-algebra K, the basechange of f by the
+identity map K → K -/
+def baseChange (f : A →ₐ[k] B) (K : Type) [CommRing K] [Algebra k K] :
+    TensorProduct k K A →ₐ[K] TensorProduct k K B :=
+    Algebra.TensorProduct.map (AlgHom.id K K) f
+
+
+lemma baseChange_by_field_inj (f : A →ₐ[k] B) (h : Function.Injective f) (K : Type) [Field K]
+    [Algebra k K] :
+    Function.Injective (Algebra.TensorProduct.map (AlgHom.id K K) f) := by
+  apply Module.Flat.lTensor_preserves_injective_linearMap
+  exact h
+
+
+/-- The k-algebra A is geometrically reduced iff its basechange to AlgebraicClosure k is reduced -/
+@[mk_iff]
+class IsGeometricallyReduced : Prop where
+  baseChangeReduced : IsReduced (TensorProduct k (AlgebraicClosure k) A)
+
+theorem IsGeometricallyReduced_imp_baseChange_by_closure_Reduced (h : IsGeometricallyReduced k A) :
+    IsReduced (TensorProduct k (AlgebraicClosure k) A) := by
+    exact h.baseChangeReduced
+
+lemma isGeometricallyReduced_of_injective (B : Type) [CommRing B] [Algebra k B] (f : A →ₐ[k] B)
+    (hf : Function.Injective f) [IsGeometricallyReduced k B] : IsGeometricallyReduced k A := by
+  let fK := baseChange  k A B f (AlgebraicClosure k)
+  have hfK : Function.Injective fK := baseChange_by_field_inj k A _ f hf _
+  expose_names
+  rw [isGeometricallyReduced_iff] at *
+  exact isReduced_of_injective fK hfK
+
+
+theorem isGeometricallyReduced_isReduced [IsGeometricallyReduced k A] : IsReduced A := by
+  -- We prove this by providing an injection A →ₐ[k] A ⊗[k] AlgebraicClosure k and then
+  -- applying isReduced_of_injective
+  let f : A →ₐ[k] TensorProduct k (AlgebraicClosure k) A := Algebra.TensorProduct.includeRight
+  expose_names
+  rw [isGeometricallyReduced_iff] at inst_3
+  have hf : Function.Injective f := by
+    apply Algebra.TensorProduct.includeRight_injective
+    exact FaithfulSMul.algebraMap_injective k (AlgebraicClosure k)
+  exact isReduced_of_injective f hf
+
+lemma notReduced_has_nilpotent {R : Type} [Zero R] [Pow R ℕ] (h : ¬IsReduced R) :
+    ∃ x : R, x ≠ 0 ∧ IsNilpotent x := by
+  by_contra h_contra
+  simp only [ne_eq, not_exists, not_and] at h_contra
+  simp only [isReduced_iff, not_forall] at h
+  obtain ⟨x, ⟨hNil, hx⟩⟩ := h
+  tauto
+
+/-- Given a subalgebra C of a k-algebra A, and a k-algebra B, the basechange of C to a subalgebra
+of A ⊗[k] B -/
+def subAlgebraBaseChange (C : Subalgebra k A) : Subalgebra B (TensorProduct k B A) :=
```

Comment:
```suggestion
def Subalgebra.baseChange (C : Subalgebra k A) : Subalgebra B (TensorProduct k B A) :=
```

Reply:
This allows you to use dot notation, i.e. you can use `s.baseChange B` to base change a sub-algebra `s` to `B`.